lucene/analysis/common/src/java/org/apache/lucene/analysis/synonym/SynonymGraphFilter.java

```diff
@@ -178,201 +157,136 @@ public SynonymGraphFilter(TokenStream input, SynonymMap synonyms, boolean ignore
     this.fstReader = fst.getBytesReader();
     scratchArc = new FST.Arc<>();
     this.ignoreCase = ignoreCase;
+    this.tokenQueue = new ArrayDeque<>();
+    this.lookahead = new ArrayDeque<>();
   }
 
   @Override
   public boolean incrementToken() throws IOException {
-    //System.out.println("\nS: incrToken lastNodeOut=" + lastNodeOut + " nextNodeOut=" + nextNodeOut);
-
-    assert lastNodeOut <= nextNodeOut;
-      
-    if (outputBuffer.isEmpty() == false) {
-      // We still have pending outputs from a prior synonym match:
-      releaseBufferedToken();
-      //System.out.println("  syn: ret buffered=" + this);
-      assert liveToken == false;
-      return true;
-    }
-
-    // Try to parse a new synonym match at the current token:
-
-    if (parse()) {
-      // A new match was found:
-      releaseBufferedToken();
-      //System.out.println("  syn: after parse, ret buffered=" + this);
-      assert liveToken == false;
-      return true;
-    }
-
-    if (lookaheadNextRead == lookaheadNextWrite) {
-
-      // Fast path: parse pulled one token, but it didn't match
-      // the start for any synonym, so we now return it "live" w/o having
-      // cloned all of its atts:
-      if (finished) {
-        //System.out.println("  syn: ret END");
+      if (!tokenQueue.isEmpty()){
```

Comment:
Can you use `== false` instead, for better future readability?  And also add a space before `{`?